fdbclient/NativeAPI.actor.cpp

```diff
@@ -3898,6 +3927,56 @@ Future<Void> getWatchFuture(Database cx, Reference<WatchParameters> parameters)
 	return Void();
 }
 
+namespace {
+
+// NOTE: Since an ACTOR could receive multiple exceptions for a single catch clause, e.g. broken promise together with
+// operation cancelled, If the decreaseWatchRefCount is placed at the catch clause, it might be triggered for multiple
+// times. One could check if the SAV isSet, but seems a more intuitive way is to use RAII-style constructor/destructor
+// pair. Yet the object has to be constructed after a wait statement, so it must be trivially-constructible. This
+// requires move-assignment operator implemented.
+class WatchRefCountUpdater {
+	Database cx;
+	int64_t tenantID;
+	KeyRef key;
+
+	void tryAddRefCount() {
+		if (cx.getReference()) {
+			cx->increaseWatchRefCount(tenantID, key);
+		}
+	}
+
+	void tryDelRefCount() {
+		if (cx.getReference()) {
+			cx->decreaseWatchRefCount(tenantID, key);
+		}
+	}
+
+public:
+	WatchRefCountUpdater() {}
+
+	WatchRefCountUpdater(const Database& cx_, const int64_t tenantID_, KeyRef key_)
+	  : cx(cx_), tenantID(tenantID_), key(key_) {
+		tryAddRefCount();
+	}
+
+	WatchRefCountUpdater& operator=(WatchRefCountUpdater&& other) {
+		tryDelRefCount();
+
+		// NOTE: Do not use move semantic, this copy allows other delete the reference count properly.
```

Comment:
Could we use move while clearing out `other.cx`? Then we shouldn't need to `tryAddRefCount` at all.